src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -714,11 +962,17 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
   if (_heap->is_concurrent_weak_root_in_progress() && r->is_trash()) {
     return nullptr;
   }
+  HeapWord* result = nullptr;
+  bool is_generational = _heap->mode()->is_generational();
+  try_recycle_trashed(r, is_generational);
+  in_new_region = r->is_empty();
 
-  try_recycle_trashed(r);
-  if (!r->is_affiliated()) {
-    ShenandoahMarkingContext* const ctx = _heap->complete_marking_context();
+  if (in_new_region) {
+    log_debug(gc)("Using new region (" SIZE_FORMAT ") for %s (" PTR_FORMAT ").",
+                       r->index(), ShenandoahAllocRequest::alloc_type_to_string(req.type()), p2i(&req));
+    assert(!r->is_affiliated(), "New region " SIZE_FORMAT " should be unaffiliated", r->index());
     r->set_affiliation(req.affiliation());
+    ShenandoahMarkingContext* const ctx = _heap->complete_marking_context();
```

Comment:
Can we move this `ctx` closer to the `asserts` that use it, or perhaps inline the variable in the asserts?

Reply:
Thanks.  Done.